pkg/watch/health.go

```diff
@@ -0,0 +1,257 @@
+package watch
+
+import (
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/square/p2/pkg/health"
+	"github.com/square/p2/pkg/kp"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/preparer"
+	"github.com/square/p2/pkg/util/net"
+)
+
+// These constants should probably all be something the p2 user can set
+// in their preparer config...
+
+// number of milliseconds between reality store checks
+const POLL_KV_FOR_PODS = 2000
+
+// number of milliseconds between health checks
+const HEALTHCHECK_INTERVAL = 1000
+
+// TTL on healthchecks in seconds
+const TTL = 60
+
+// Contains method for watching the consul reality store to
+// track services running on a node. A manager method:
+// WatchHealth tracks the reality store and manages
+// a health checking go routine for each service in the
+// reality store
+
+// PodWatch houses a pod's manifest, a channel to kill the
+// pod's goroutine if the pod is removed from the reality
+// tree, and a bool that indicates whether or not the pod
+// has a running MonitorHealth go routine
+type PodWatch struct {
+	manifest kp.ManifestResult
+
+	// For tracking/controlling the go routine that performs health checks
+	// on the pod associated with this PodWatch
+	shutdownCh chan bool
+	hasMonitor bool // indicates whether this pod is being monitored
+
+	// the fields are provided so it can be determined if new health checks
+	// actually need to be sent to consul. If newT - oldT << TTL and status
+	// has not changed there is no reason to update consul
+	lastCheck  time.Time          // time of last health check
+	lastStatus health.HealthState // status of last health check
+
+	logger *logging.Logger
+}
+
+// WatchHealth is meant to be a long running go routine.
+// WatchHealth reads from a consul store to determine which
+// services should be running on the host. WatchHealth
+// runs a CheckHealth routine to monitor the health of each
+// service and kills routines for services that should no
+// longer be running.
+func WatchHealth(config *preparer.PreparerConfig, logger *logging.Logger, shutdownCh chan struct{}) {
```

Comment:
I think this name is a little confusing since you're not using a watch operation, and you're not watching consul for data like the other Watch*. All that said, I don't have a better name :). Maybe SyncAndMonitorPods?